### PR TITLE
Handle Kubernetes node event and cond.

### PR DESCRIPTION
**Pull request by mantissahz**
Move checking the manager pod status after checking Kubernetes node status and not check the manager pod status if Kubernetes node status is not ready.

longhorn/longhorn#1449

### CODE REVIEW

File: controller/node_controller.go [hunk 1]
```diff
@@ -435,6 +399,52 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
+	nodeManagerFound := false
+	for _, pod := range managerPods {
+		if pod.Spec.NodeName != node.Name {
+			continue
+		}
+		nodeManagerFound = true
+		if !kubeNodeIsReady {
+			break
+		}
+		podConditions := pod.Status.Conditions
```
Comment (innobead) [inline controller/node_controller.go:416]: Can you format this way?

```go
if pod.Spec.NodeName != node.Name {
    continue
}

nodeManagerFound = true
if !kubeNodeIsReady {
    break
}

podConditions := pod.Status.Conditions
...
```

File: controller/node_controller.go [hunk 2]
```diff
@@ -435,6 +399,52 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
+	}
+	nodeManagerFound := false
+	for _, pod := range managerPods {
+		if pod.Spec.NodeName != node.Name {
+			continue
+		}
+		nodeManagerFound = true
+		if !kubeNodeIsReady {
+			break
+		}
```
Comment (innobead) [inline controller/node_controller.go:415]: Don't think you need this, but rather let the below logic to update the node condition.

File: controller/node_controller.go [hunk 3]
```diff
@@ -435,6 +399,52 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
+				}
+				node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady, longhorn.ConditionStatusTrue, "", nodeIsReadyEventMsg)
+			} else {
+				// create the event after `UpdateNodeStatus` without the error `IsConflict`
+				if types.GetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady).Status != longhorn.ConditionStatusFalse {
+					nodeIsReadyEventMsg = fmt.Sprintf("Node %v is down: the manager pod %v is not running", node.Name, pod.Name)
+					nodeIsReadyEventType = v1.EventTypeWarning
+				}
+				node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady, longhorn.ConditionStatusFalse, string(longhorn.NodeConditionReasonManagerPodDown), nodeIsReadyEventMsg)
+			}
```
Comment (innobead) [inline controller/node_controller.go:435]: Why not?

```go
if podCondition.Status == v1.ConditionTrue && pod.Status.Phase == v1.PodRunning {
    // create the event after `UpdateNodeStatus` without the error `IsConflict`
    if types.GetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady).Status != longhorn.ConditionStatusTrue {
        nodeIsReadyEventMsg = fmt.Sprintf("Node %v is ready", node.Name)
        nodeIsReadyEventType = v1.EventTypeNormal
        node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady, longhorn.ConditionStatusTrue, "", nodeIsReadyEventMsg)    
    }
} else {
    // create the event after `UpdateNodeStatus` without the error `IsConflict`
    if types.GetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady).Status != longhorn.ConditionStatusFalse {
        nodeIsReadyEventMsg = fmt.Sprintf("Node %v is down: the manager pod %v is not running", node.Name, pod.Name)
        nodeIsReadyEventType = v1.EventTypeWarning
        node.Status.Conditions = types.SetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady, longhorn.ConditionStatusFalse, string(longhorn.NodeConditionReasonManagerPodDown), nodeIsReadyEventMsg)
    }
}
```

**Review by innobead — CHANGES_REQUESTED** [review]

File: controller/node_controller.go (hunk 1 above)
Comment (mantissahz) [reply controller/node_controller.go]: Modified.

File: controller/node_controller.go [hunk 4]
```diff
@@ -435,6 +399,52 @@ func (nc *NodeController) syncNode(key string) (err error) {
 
 	}
 
+	// sync node state by manager pod
+	managerPods, err := nc.ds.ListManagerPods()
+	if err != nil {
+		return err
+	}
+	nodeManagerFound := false
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply controller/node_controller.go]: Modified.

**Comment by innobead** [comment]
ref: https://github.com/longhorn/longhorn/issues/1449#issuecomment-1319574678

**Comment by mantissahz** [comment]
According to https://github.com/longhorn/longhorn/issues/1449#issuecomment-1319574678
Close this PR.

**Pull request closed without merging**